lib/Dialect/Polynomial/IR/PolynomialCanonicalization.td

```diff
@@ -1,25 +1,54 @@
 #ifndef LIB_DIALECT_POLYNOMIAL_IR_POLYNOMIALCANONICALIZATION_TD_
 #define LIB_DIALECT_POLYNOMIAL_IR_POLYNOMIALCANONICALIZATION_TD_
 
+include "lib/Dialect/ModArith/IR/ModArithOps.td"
 include "lib/Dialect/Polynomial/IR/PolynomialOps.td"
 include "mlir/Dialect/Arith/IR/ArithOps.td"
 include "mlir/IR/OpBase.td"
 include "mlir/IR/PatternBase.td"
 
 def Equal : Constraint<CPred<"$0 == $1">>;
 
+def isIntegerCoeffType
+  : Constraint<CPred<
+        "isa<IntegerType>(cast<PolynomialType>($0.getType()).getRing().getCoefficientType())">,
+  "coefficients have integer type">;
+
+def isModArithCoeffType
+  : Constraint<CPred<
+        "isa<mod_arith::ModArithType>(cast<PolynomialType>($0.getType()).getRing().getCoefficientType())">,
+  "coefficients have mod_arith type">;
+
 // Get a -1 integer attribute of the same type as the polynomial SSA value's
 // ring coefficient type.
 def getMinusOne
   : NativeCodeCall<
       "$_builder.getIntegerAttr("
         "cast<PolynomialType>($0.getType()).getRing().getCoefficientType(), -1)">;
 
-def SubAsAdd : Pat<
+def getMinusOneModArith
+  : NativeCodeCall<
+      "mod_arith::ModArithAttr::get("
+        "cast<mod_arith::ModArithType>("
+          "cast<PolynomialType>("
+            "$0.getType()).getRing().getCoefficientType()"
+          "), -1)">;
+
+def SubAsAddInt : Pat<
+  (Polynomial_SubOp $f, $g),
+  (Polynomial_AddOp $f,
+    (Polynomial_MulScalarOp $g,
+      (Arith_ConstantOp (getMinusOne $g)))),
+  [(isIntegerCoeffType $g)]
+>;
+
+def SubAsAddModArith : Pat<
   (Polynomial_SubOp $f, $g),
   (Polynomial_AddOp $f,
     (Polynomial_MulScalarOp $g,
-      (Arith_ConstantOp (getMinusOne $g))))>;
+        (ModArith_ConstantOp (getMinusOneModArith $g)))),
+  [(isModArithCoeffType $g)]
+>;
```

Comment:
I think `sub = add (mulScalar)` is costly as `modarith.sub` is much cheaper than `modarith.add (modarith.mul).` As now the `ConvertAdd` in `PolynomialToStandard` is simple enough, we can have a corresponding `ConvertSub`

Reply:
Removed!